### PR TITLE
Update search logic to include blog name and status

### LINKED ISSUE #19484: 🤖 Update search logic to include blog name and status
This task is to update the "My domains" search logic to match on `blogName` and `status`.

### PR DESCRIPTION
**Pull request by justtwago**
Fixes #19484

This PR includes additional "blog name" and "status" properties to filter my domains list by. In addition, I moved the filter logic from the UI layer to the ViewModel to make it more efficient. The previous implementation caused unnecessary invoking of filtering while scrolling the list. Composables should just render data to avoid performance issues caused by repeated calculations while recomposition.

To test:
- Make sure you enabled `domain_management` feature flag
- If your account doesn't have a list of domains you can use a mocked list via sandboxing.  (Details here - p1692765970283879-slack-C05NS0YV7HS) Unfortunately, you can only search by domain and status using the mocked domains.
- Go to Me -> Domains
- [ ] Make sure you can search locally for domains by blog name and domain status

## Regression Notes
1. Potential unintended areas of impact
All Domains screen

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

### CODE REVIEW

**Comment by justtwago** [comment]
I'll make it ready for review once add all missing tests for `uiStateFlow` changes and search logic.

File: WordPress/src/main/java/org/wordpress/android/ui/domains/management/MyDomainsScreen.kt [hunk 1]
```diff
@@ -87,12 +88,15 @@ fun MyDomainsScreen(
...
+                onQueryStringChanged = {
+                    queryString = it
+                    onSearchQueryChanged(it)
+                },
+                enabled = uiState is PopulatedList.Loaded || uiState is PopulatedList.Filtered,
             )
             when (uiState) {
                 is PopulatedList -> MyDomainsList(
-                    listUiState = uiState.filter(queryString),
+                    listUiState = uiState,
```
Comment (mkevins) [inline WordPress/src/main/java/org/wordpress/android/ui/domains/management/MyDomainsScreen.kt:99]: Nice catch, and good idea to stabilize this for performance 👍 .

File: WordPress/src/main/java/org/wordpress/android/ui/domains/management/MyDomainsScreen.kt [hunk 2]
```diff
@@ -87,12 +88,15 @@ fun MyDomainsScreen(
             MyDomainsSearchInput(
                 elevation.value,
                 queryString = queryString,
-                onQueryStringChanged = { queryString = it },
-                enabled = uiState is PopulatedList.Loaded,
+                onQueryStringChanged = {
+                    queryString = it
+                    onSearchQueryChanged(it)
+                },
+                enabled = uiState is PopulatedList.Loaded || uiState is PopulatedList.Filtered,
```
Comment (mkevins) [inline WordPress/src/main/java/org/wordpress/android/ui/domains/management/MyDomainsScreen.kt]: Please don't consider this a blocker, but as a suggestion, it feels like `Filtered` is also `Loaded`, so maybe we can have it modeled as such? Then we wouldn't need this extra logic. 🤔

File: WordPress/src/main/java/org/wordpress/android/ui/domains/management/util/DomainLocalSearchEngine.kt [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package org.wordpress.android.ui.domains.management.util
+
+import org.wordpress.android.fluxc.network.rest.wpcom.site.AllDomainsDomain
+import javax.inject.Inject
+
+class DomainLocalSearchEngine @Inject constructor() {
```
Comment (mkevins) [inline WordPress/src/main/java/org/wordpress/android/ui/domains/management/util/DomainLocalSearchEngine.kt:6]: Good idea to extract this, and also, nice name 😄

File: WordPress/src/test/java/org/wordpress/android/ui/domains/management/util/DomainLocalSearchEngineTest.kt [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        val query = "chocolate"
+
+        // When
+        val result = engine.filter(allDomains, query)
+
+        // Then
+        assertThat(result).isEqualTo(listOf(fakeDomainBar))
+    }
+
+    @Test
```
Comment (mkevins) [inline WordPress/src/test/java/org/wordpress/android/ui/domains/management/util/DomainLocalSearchEngineTest.kt:55]: Thank you for adding the extra tests!

**Review by mkevins — COMMENTED** [review]
I gave this a look, though it is still marked as draft, and the approach looks good. I made one minor comment about the possibility of restructuring the data / sealed classes for ui state, but it is non-blocking.

File: WordPress/src/main/java/org/wordpress/android/ui/domains/management/MyDomainsScreen.kt (hunk 2 above)
Comment (justtwago) [reply WordPress/src/main/java/org/wordpress/android/ui/domains/management/MyDomainsScreen.kt]: Yeah, 100% agree! Thanks for the suggestion!
**Upd:** Pushed an improvement for this.

**Comment by justtwago** [comment]
FYI, I played with an advanced search algorithm to handle queries with typos, but it occurred it wasn't efficient for smaller queries, so decided to keep it simple. If you are curious about Levenshtein distance, I saved two first commits with the POC.

**Review by antonis — APPROVED** [review]
Great work @justtwago 🥇 
I've tested the implementation with the latest changes and everything worked for me as expected. The code also looks good 🎉 
Thank you also for adding extensive test 🙇

**Pull request merged**